### PR TITLE
Hyper-V : Added 4 Network Gen2 VM tests

### PR DESCRIPTION
**Pull request by asofron**
VHD under test **ubuntu_18.04_x64_gen2vm.vhdx**
 ARM Image under test  :  :  : 
 Total Executed TestCases        4 (       4 Pass,        0 Fail,        0 Abort)
 Total Execution Time(dd:hh:mm) 0:0:9
 XML file: TestConfiguration
 
 #Sr. Test Name : Test Result : Test Duration (in minutes)
 ----------------------------------------------------
 1. NET-HOT-ADD-MULTINIC : PASS : 2.15
 2. NET-HOT-REMOVE-MULTINIC : PASS : 0.65
 3. NET-BOOT-NONIC-HOT-ADD-NIC : PASS : 2.33
 4. NET-HOT-ADD-REMOVE-MAXNIC : PASS : 4.07

### CODE REVIEW

File: Libraries/CommonFunctions.psm1 [hunk 1]
```diff
@@ -4284,4 +4284,104 @@ Function Remove-InvalidCharactersFromFileName
...
+
+function Get-KVPItem {
+    param (
+        $VMName,
+        $server,
+        $keyName,
+        $Intrinsic
+    )
+
+    $vm = Get-WmiObject -Namespace root\virtualization\v2 -Query "Select * From Msvm_ComputerSystem Where ElementName=`'$VMName`'"
```
Comment (chvalean) [inline Libraries/CommonFunctions.psm1]: missing ComputerName variable

File: Libraries/CommonFunctions.psm1 [hunk 2]
```diff
@@ -4284,4 +4284,104 @@ Function Remove-InvalidCharactersFromFileName
...
+    if (-not $vm)
+    {
+        return $Null
```
Comment (chvalean) [inline Libraries/CommonFunctions.psm1]: please confirm that if this is hit, the test execution will end

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    LogMsg "Info : Checking the eth_count"
+    if [ $eth_count -ne 2 ]; then
+        LogErr "VM should have two NICs now"
+        SetTestStateAborted
+        exit 0
+    fi
+    # Bring the new NIC online
+    os_vendor=$(lsb_release -i -s)
+    LogMsg "os_vendor=$os_vendor"
+    if [[ "$os_vendor" == "Red Hat" ]] || \
```
Comment (mbivolan) [inline Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: You can use the OS_FAMILY env variable here.
https://github.com/LIS/LISAv2/blob/master/Testscripts/Linux/utils.sh#L327

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+         [ "$os_vendor" == "Debian" ]; then
+            echo "auto ${eth_name}" >> /etc/network/interfaces
+            echo "iface ${eth_name} inet dhcp" >> /etc/network/interfaces
+    else
+        LogErr "Linux Distro not supported!"
+        SetTestStateAborted
+        exit 0
+    fi
+
+    LogMsg "Info : Bringing up ${eth_name}"
```
Comment (mbivolan) [inline Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: Remove "Info :"

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    eth_name=$2
+    if [ $eth_count -ne 1 ]; then
+        LogErr "There are more than one eth devices"
+        SetTestStateAborted
+        exit 0
+    fi
+    # Clean up files & check linux log for errors
+    os_vendor=$(lsb_release -i -s)
+    LogMsg "os_vendor=$os_vendor"
+    if [[ "$os_vendor" == "Red Hat" ]] || \
```
Comment (mbivolan) [inline Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: OS_FAMILY

File: Testscripts/Linux/NET-Verify-Boot-NoNIC.sh [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    sleep 10
+    if [ $timeout -le 0 ]; then
+        echo "Timed out waiting for eth device to be created"
+        exit 0
+    fi
+done
+
+# Bring up the new eth device
+ifup eth0
+
```
Comment (chvalean) [inline Testscripts/Linux/NET-Verify-Boot-NoNIC.sh]: add a sleep here, to allow some time for dhcpd to set an IP on the interface and get connectivity

File: XML/TestCases/FunctionalTests.xml [hunk 7]
```diff
@@ -1370,4 +1370,53 @@
...
+        <files>.\Testscripts\Linux\NET-MAX-NICs.sh,.\Testscripts\Linux\NET-Verify-HotAdd-MultiNIC.sh,.\Testscripts\Linux\utils.sh</files>
+        <setupType>SingleVM</setupType>
+        <TestParameters>
+            <param>TEST_TYPE=synthetic</param>
+            <param>SYNTHETIC_NICS=7</param>
+        </TestParameters>
+        <Platform>HyperV</Platform>
+        <Category>Functional</Category>
+        <Area>NET</Area>
+        <Tags>net</Tags>
```
Comment (mbivolan) [inline XML/TestCases/FunctionalTests.xml]: add closing "test" tag here

File: Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1 [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    RemoteCopy -download -downloadFrom $Ipv4 -files "/home/${VMUserName}/NETVerifyHotRemoveMultiNIC.log" `
+        -downloadTo $LogDir -port $VMPort -username $VMUserName -password $VMPassword
+    $contents = Get-Content -Path $stateFile
+    $contents2 = Get-Content -Path $logFile
+    if (($contents -eq "TestAborted") -or ($contents -eq "TestFailed")) {
+        LogErr "Error: Running $remoteScript script failed on VM!"
+        return "FAIL"
+    }
+    if ($contents2 -match "netvsc throwed errors") {
+        LogErr "Error: VM '${vmName}' reported that netvsc throwed errors"
```
Comment (mbivolan) [inline Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1]: Remove "Error: "

File: Testscripts/Linux/NET-Verify-Boot-NoNIC.sh (hunk 6 above)
Comment (asofron) [reply Testscripts/Linux/NET-Verify-Boot-NoNIC.sh]: Added sleep

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh (hunk 3 above)
Comment (asofron) [reply Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: Added OS_FAMILY

File: XML/TestCases/FunctionalTests.xml (hunk 7 above)
Comment (asofron) [reply XML/TestCases/FunctionalTests.xml]: Fixed

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh (hunk 4 above)
Comment (asofron) [reply Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: Removed

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh (hunk 5 above)
Comment (asofron) [reply Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: Changed to OS_FAMILY

File: Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1 (hunk 8 above)
Comment (asofron) [reply Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1]: Removed

File: Libraries/CommonFunctions.psm1 [hunk 9]
```diff
@@ -4284,4 +4284,104 @@ Function Remove-InvalidCharactersFromFileName
...
+        if ($key -eq $keyName)
+        {
+            return $value
+        }
+    }
+
+    return $Null
+}
+
+function Add-RemoveMaxNIC {
```
Comment (juhlee-microsoft) [inline Libraries/CommonFunctions.psm1]: the name sounds confused.

File: Testscripts/Windows/NET-Boot-NoNIC-HotAdd-NIC.ps1 [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+        RunLinuxCmd -username $VMUserName -password $VMPassword -ip $Ipv4 -port $VMPort -command `
+            "echo 'chmod 755 /home/$VMUserName/NET-Verify-Boot-NoNIC.sh' >> /etc/rc.d/rc.local" -runAsSudo
+        RunLinuxCmd -username $VMUserName -password $VMPassword -ip $Ipv4 -port $VMPort -command `
+            "echo 'cd /home/$VMUserName/ && ./NET-Verify-Boot-NoNIC.sh > /home/$VMUserName/NET-Verify-Boot-NoNIC.log  &' >> /etc/rc.d/rc.local" -runAsSudo
+    }
+    elseif ($linuxRelease -eq "UBUNTU" -or $linuxRelease -eq "DEBIAN") {
+        RunLinuxCmd -username $VMUserName -password $VMPassword -ip $Ipv4 -port $VMPort -command `
+            "systemctl start cron.service" -runAsSudo
+        RunLinuxCmd -username $VMUserName  -password $VMPassword -ip $Ipv4 -port $VMPort -command `
+            "echo '@reboot chmod 755 /home/$VMUserName/NET-Verify-Boot-NoNIC.sh' >> /var/spool/cron/crontabs/root"  -runAsSudo
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/NET-Boot-NoNIC-HotAdd-NIC.ps1]: Can you consider if this is running under root?

File: Testscripts/Linux/NET-Verify-Boot-NoNIC.sh [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+ip addr show eth0 | grep "inet\b"
+if [ $? -ne 0 ]; then
+    echo "eth0 was not assigned an IPv4 address"
+    exit 0
+fi
+
+echo "Info : eth0 is up"
+
+# Modify the KVP HotAddTest value to 'NICUp'
+echo "Info : Updating HotAddTesk KVP item to 'NICUp'"
```
Comment (lizzha) [inline Testscripts/Linux/NET-Verify-Boot-NoNIC.sh]: Typo: HotAddTesk

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+    fi
+
+    LogMsg "Bringing up ${eth_name}"
+    ifup ${eth_name}
+    sleep 5
+    # Verify the new NIC received an IP v4 address
+    LogMsg "Verify the new NIC has an IPv4 address"
+    #ifconfig ${eth_name} | grep -s "inet " > /dev/null
+    ip addr show ${eth_name} | grep "inet\b" > /dev/null
+    if [ $? -ne 0 ]; then
```
Comment (lizzha) [inline Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: use check_exit_status instead

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    echo "Error: unable to source utils.sh!"
+    exit 0
+}
+
+# Source constants file and initialize most common variables
+UtilsInit
+
+
+# Added_Nic ($eth_count)
+function Added_Nic {
```
Comment (lizzha) [inline Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: name it Add_Nic?

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        LogErr "${eth_name} was not assigned an IPv4 address"
+        SetTestStateAborted
+        exit 0
+    fi
+    LogMsg "${eth_name} is up"
+    LogMsg "NIC Hot Add test passed"
+}
+
+# Removed_Nic ($eth_count)
+function Removed_Nic {
```
Comment (lizzha) [inline Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: name it Remove_Nic?

File: Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1 [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+        $tokens = $p.Trim().Split("=")
+        if ($tokens.Length -ne 2)
+        {
+            continue
+        }
+        $val = $tokens[1].Trim()
+        switch($tokens[0].Trim().ToLower())
+        {
+		"SYNTHETIC_NICS"{ $nicsAmount  = $val -as [int] }
+        default         { continue }
```
Comment (lizzha) [inline Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1]: fix the indent

File: Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1 [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    #	Hot Add maximum number of synthetic NICs
+    LogMsg "Hot Adding the maximum number of synthetic NICs ..."
+    $addnic = Add-RemoveMaxNIC $vmName $hvServer $switchName "add" $nicsAmount
+
+    # Run the NET_MAX_NIC.sh on the SUT VM to verify the VM detected the hot add
+    LogMsg "Verifing the OS detected the NIC was hot add..."
+
+    $stateFile = "${LogDir}\state.txt"
```
Comment (lizzha) [inline Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1]: this part can be replaced with the function added in PR 307
https://github.com/LIS/LISAv2/blob/56aa2f3f2c2eaef11c5485c050afb70d40a48345/Libraries/CommonFunctions.psm1#L4291

File: Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1 [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+        }
+    }
+
+    # Now Hot Remove the NIC
+    $removenic = Add-RemoveMaxNIC  $vmName  $hvServer  $switchName "remove" $nicsAmount
+
+    # Run the NET_VerifyHotAddSyntheticNIC.sh on the SUT VM to verify the VM detected the hot remove
+    LogMsg "Verifing the OS detected the NIC was hot removed..."
+    $action = "remove"
+    $stateFile = "${LogDir}\state.txt"
```
Comment (lizzha) [inline Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1]: this part can be replaced with the function added in PR 307
https://github.com/LIS/LISAv2/blob/56aa2f3f2c2eaef11c5485c050afb70d40a48345/Libraries/CommonFunctions.psm1#L4291

File: Libraries/CommonFunctions.psm1 (hunk 9 above)
Comment (asofron) [reply Libraries/CommonFunctions.psm1]: Replaced with Test-MaxNIC

File: Testscripts/Windows/NET-Boot-NoNIC-HotAdd-NIC.ps1 (hunk 10 above)
Comment (asofron) [reply Testscripts/Windows/NET-Boot-NoNIC-HotAdd-NIC.ps1]: Yes root is needed

File: Testscripts/Linux/NET-Verify-Boot-NoNIC.sh (hunk 11 above)
Comment (asofron) [reply Testscripts/Linux/NET-Verify-Boot-NoNIC.sh]: Fixed

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh (hunk 12 above)
Comment (asofron) [reply Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: replaced with check_exit_status

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh (hunk 13 above)
Comment (asofron) [reply Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: Renamed

File: Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh (hunk 14 above)
Comment (asofron) [reply Testscripts/Linux/NET-Verify-HotAdd-MultiNIC.sh]: Renamed

File: Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1 (hunk 15 above)
Comment (asofron) [reply Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1]: Fixed

File: Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1 (hunk 16 above)
Comment (asofron) [reply Testscripts/Windows/NET-HotAddRemove-Max-NIC.ps1]: Does not apply and PR 307 is not merge

File: Testscripts/Windows/NET-Boot-NoNIC-HotAdd-NIC.ps1 (hunk 10 above)
Comment (adriansuhov) [reply Testscripts/Windows/NET-Boot-NoNIC-HotAdd-NIC.ps1]: @juhlee-microsoft Please consider that /home/$user is hardcoded in a lot of places in the framework. So if root user is given in the testConfiguration.xml as the HyperV user, the framework itself will fail, before actually getting here

Comment (santoshx) [reply Testscripts/Windows/NET-Boot-NoNIC-HotAdd-NIC.ps1]: May be env variable $HOME can be used.

**Comment by chvalean** [comment]
@juhlee-microsoft @lizzha I think all comments have been addressed, please review and sign-off.
Thank you!

File: Testscripts/Windows/NET-Boot-NoNIC-HotAdd-NIC.ps1 (hunk 10 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/NET-Boot-NoNIC-HotAdd-NIC.ps1]: acked

**Pull request merged**